### PR TITLE
Adds a Grafana dashboard for the metrics gathered in Prometheus

### PR DESCRIPTION
**Pull request by josemigallas**
## Description
Adds a fancy dashboard that will be displayed in Grafana, provided a metrics service is provisioned.

JIRA: https://issues.jboss.org/browse/AEROGEAR-1938

### Note
Pie Charts are not included in Grafana by default, they have to be installed as a plugin and unfortunately there's no straightforward way to do this.
@darahayes has been investigating this and wrote a document with some suggestions.

Currently the only way is doing it from the pod itself and then restart it. In `grafana` terminal:
```sh
$ grafana-cli plugins install grafana-piechart-panel
```
### Verification
For this to work:

1. Metrics has to be provisioned before Keycloak
2. Grafana's pod has to be restarted once Keycloak is provisioned.
3. Pie-Chart plugin has to be installed manually for these charts to be shown.
4. To make the dashboard editable, navigate to dashboard settings (cog image at the top) and click on "Make Editable".
5. Expect to see the dashboard empty but without any warnings (left-top corner of each panel)

### Tasks
- [x] Fancy dashboard
- [x] Useful metrics
- [x] Automatically detect different identity providers
- [x] Automatically added to Grafana once metrics is provisioned
- [ ] Display pie-charts (install plugin needed)

### CODE REVIEW

**Comment by darahayes** [comment]
Thanks @josemigallas just to add more info about the [document](https://docs.google.com/document/d/1WAUQ9IfJovSSUsHYR2PU3UeRWdpuyoKN9_ImyVv6dhc/edit?usp=sharing) I wrote. This is the current state of plugins in Grafana:

Grafana Plugins are installed using 1 of two ways

- Drop plugin repo into a configured plugins directory
- CLI tool* - grafana-cli plugins install <plugin-name>

*official plugins only, it basically performs 1st method for you, installing plugins requires a restart of Grafana.

The nature of Grafana plugins presents challenges for getting them working in OpenShift. Here are some suggestions for how we can get Grafana plugins in OpenShift. None are ideal IMO.

1. Store Plugin repos in a Persistent Volume which can be mounted to 1 or more Pods.
	- Docker image/Dockerfile does not need to be modified
	- Easy to maintain in metrics-apb
	- Not sure how well this would work in a production environment like AWS
2. Store Plugin repos in a Config Map.
	- Docker image/Dockerfile does not need to be modified
	- Probably not best practice? Probably limitations in Config Maps.
3. Override Start command to install a list of plugins before starting Grafana
	- Easy implementation, easy to maintain
	- Probably unsuitable for many customer environments (network restrictions)
4. Create our own Docker Image with plugins baked in
	- Relatively easy (for now)
	- Need to maintain our own image
	- Bakes in plugins without options for user to add or remove as they see fit

I think we'll eventually have to go down the route of option 4 but I'm not really sure. @david-martin  any thoughts on this? I think we need to open a JIRA ticket for this.

File: Dockerfile [hunk 1]
```diff
@@ -2,22 +2,25 @@ FROM ansibleplaybookbundle/apb-base
...
-YW1lOiBkZWZhdWx0CiAgICBkZXNjcmlwdGlvbjogRGVwbG95IGtleWNsb2FrCiAgICBmcmVlOiBU\
-cnVlCiAgICBtZXRhZGF0YToge30KICAgIHBhcmFtZXRlcnM6IAogICAgLSBuYW1lOiBBRE1JTl9O\
-QU1FCiAgICAgIHJlcXVpcmVkOiBUcnVlCiAgICAgIGRlZmF1bHQ6IGFkbWluCiAgICAgIHR5cGU6\
-IHN0cmluZwogICAgICB0aXRsZTogS2V5Y2xvYWsgYWRtaW4gdXNlcm5hbWUKICAgIC0gbmFtZTog\
-QURNSU5fUEFTU1dPUkQKICAgICAgcmVxdWlyZWQ6IFRydWUKICAgICAgZGVmYXVsdDogYWRtaW4K\
-ICAgICAgdHlwZTogc3RyaW5nCiAgICAgIHRpdGxlOiBLZXljbG9hayBhZG1pbiBwYXNzd29yZAog\
-ICAgYmluZF9wYXJhbWV0ZXJzOgogICAgLSBuYW1lOiBzZXJ2aWNlCiAgICAgIHRpdGxlOiBUaGUg\
-c2VydmljZSB0aGUgY2xpZW50IGlzIGZvcgogICAgICB0eXBlOiBzdHJpbmcK"
+"dmVyc2lvbjogMS4wCm5hbWU6IGtleWNsb2FrLWpvc2UtYXBiCmRlc2NyaXB0aW9uOiBLZXljbG9h\
+ayAtIE9wZW4gU291cmNlIElkZW50aXR5IGFuZCBBY2Nlc3MgTWFuYWdlbWVudApiaW5kYWJsZTog\
```
Comment (darahayes) [inline Dockerfile]: This shouldn't have changed. Please remove the change

**Comment by aliok** [comment]
@josemigallas @darahayes 
Sorry guys something is not clear for me. I was able to install Grafana plugins by passing their names in an env var. 
`GF_INSTALL_PLUGINS=grafana-clock-panel,grafana-simple-json-datasource`
See https://github.com/grafana/grafana-docker#installing-plugins-for-grafana-3

Doesn't this work in this case?

**Comment by josemigallas** [comment]
@aliok Normally yes, but they way we instantiate the container [(here)](https://github.com/aerogearcatalog/metrics-apb/blob/master/roles/provision-metrics-apb/tasks/main.yml#L165-L166), Ansible overrides the default Docker entry point [(here)](https://github.com/grafana/grafana-docker/blob/master/Dockerfile#L22). For us, adding that env var does nothing at all.

**Comment by aliok** [comment]
@josemigallas ok, I would still go and try to solve this problem by making the env vars used somehow.

ping @philbrookes any ideas? Git blame tells that you're the original author :) What was the reason that you overrode the command?

**Comment by philbrookes** [comment]
@aliok I don't recall why I did it that way. Can the line be removed? Does Grafana still work without it? My best guess is that I copied that from an online example somewhere.

**Comment by david-martin** [comment]
@aliok does the `GF_INSTALL_PLUGINS` env var approach mean that grafana will pull the plugins on startup (from some external site) and install them?
If yes, that's OK to get something working.
However, we may want to use a different 'bundling' approach for security reasons in future. e.g. for a downstream/product build

**Comment by pb82** [comment]
@david-martin @aliok it looks like that's the case:

```
Pass the plugins you want installed to docker with the GF_INSTALL_PLUGINS environment variable as a comma seperated list. This will pass each plugin name to grafana-cli plugins install ${plugin}.
```

**Comment by darahayes** [comment]
@aliok @philbrookes @pb82 @josemigallas 

The container will fail to start in OpenShift using the default entrypoint

The reason the default entrypoint [`run.sh`](https://github.com/grafana/grafana-docker/blob/master/run.sh) was overridden was because the script tries to switch users using an `exec gosu` command. Such an operation requires root privileges which is not allowed in OpenShift by default.

**Comment by darahayes** [comment]
@david-martin and myself had a conversation about pushing up our own Grafana image to the aerogear dockerhub. This is a temporary solution until we have to figure out how to productise Grafana.

The reason for this is to keep the image somewhat stable. Right now we're relying on features only available in Grafana master. The master image is being updated several times a day which is undesirable.

I could temporarily solve the plugins issue at the same time by forking the Grafana docker build tools and modifying the run.sh entrypoint so it doesn't require root. This would allow us to use the `GF_INSTALL_PLUGINS` environment variable. How does that sound?

**Comment by aliok** [comment]
@darahayes that sounds good to me.
well, let's merge this PR and then do that as a separate task IMO.

**Comment by darahayes** [comment]
@aliok myself and @StevenTobin are working on a solution.

**Comment by josemigallas** [comment]
I think Dara's proposal of building our own image is a very convenient one. It would give us stability in exchange of having to build it regularly enough (which shouldn't have to be very regularly in fact).

**Comment by aliok** [comment]
@philbrookes

I see you (or the script you've found online :smile: ) overcome the problem of using "gosu" by  changing the [file modes](https://github.com/aerogearcatalog/metrics-apb/blob/master/roles/provision-metrics-apb/tasks/main.yml#L190) and then overriding the  [command](https://github.com/aerogearcatalog/metrics-apb/blob/master/roles/provision-metrics-apb/tasks/main.yml#L165). 

But, what's the proper way of making this work without overriding the command? Like, if Grafana were supporting that.
Is there something we can suggest to Grafana people?

cc @david-martin

**Comment by pb82** [comment]
@josemigallas @darahayes This PR adds the dashboard JSON, but does not update the ConfigMap for auto discovery as described in https://github.com/aerogearcatalog/metrics-apb#how-to-add-a-new-dashboard-while-provisioning-a-service

Are you going to add this in a separate PR or is there anything else preventing us to do this?

**Comment by josemigallas** [comment]
@pb82 it was previously added by @darahayes in #30

**Comment by darahayes** [comment]
@pb82 Yes, the functionality has been put in place with an empty dashboard. This PR simply adds the appropriate dashboard into it. There's no reason we shouldn't merge this PR now. It works as advertised.

In relation to the Grafana issue, myself and @StevenTobin spent time on Friday developing our own custom Grafana image that will stabilize our development and also allow us to add plugins. This image has been pushed to the aerogearcatalog organization in dockerhub. It's not production ready by any means but it will keep things going for the time being. I'll be opening a PR against the metrics-apb in the next hour or so.

**Comment by pb82** [comment]
Verified locally. Requires a restart of the grafana pod though. Tried creating a user in keycloak and logging in: worked.

bildschirmfoto 2018-02-05 um 12 19 00

**Pull request merged**